Refresh.GameServer/Middlewares/DigestMiddleware.cs

```diff
@@ -1,114 +1,158 @@
-    {
-        string url = context.Uri.AbsolutePath;
-        string auth = context.Cookies["MM_AUTH"] ?? string.Empty;
-    
-        string digestResponse = CalculateDigest(url, context.ResponseStream, auth, null, null);
+        next();
+
+        GameDatabaseContext gameDatabase = (GameDatabaseContext)database.Value;
+
+        Token? token = gameDatabase.GetTokenFromTokenData(auth, TokenType.Game);
         
-        context.ResponseHeaders["X-Digest-A"] = digestResponse;
-    }
+        // Make sure the digest calculation reads the whole response stream
+        context.ResponseStream.Seek(0, SeekOrigin.Begin);
 
-    public void HandleRequest(ListenerContext context, Lazy<IDatabaseContext> database, Action next)
-    {
-        //If this isn't an LBP endpoint, dont do digest
-        if (!context.Uri.AbsolutePath.StartsWith(GameEndpointAttribute.BaseRoute))
+        // If the digest is already saved on the token, use the token's digest
+        if (token is { Digest: not null })
         {
-            next();
+            SetDigestResponse(context, CalculateDigest(token.Digest, route, context.ResponseStream, auth, null, isUpload, token.IsHmacDigest));
             return;
         }
 
-        short? exeVersion = null;
-        short? dataVersion = null;
-        if (short.TryParse(context.RequestHeaders["X-Exe-V"], out short exeVer))
+        foreach (string digest in config.Sha1DigestKeys)
         {
-            exeVersion = exeVer;
+            string calculatedClientDigest = CalculateDigest(digest, route, context.InputStream, auth, pspVersionInfo, isUpload, false);
+
+            // If the calculated client digest is invalid, then this isn't the digest the game is using, so check the next one
+            if (calculatedClientDigest != clientDigest) 
+                continue;
+
+            SetDigestResponse(context, CalculateDigest(digest, route, context.ResponseStream, auth, null, isUpload, false));
+            
+            if(token != null)
+                gameDatabase.SetTokenDigestInfo(token, digest, false);
+            
+            return;
         }
-        if (short.TryParse(context.RequestHeaders["X-Data-V"], out short dataVer))
+
+        foreach (string digest in config.HmacDigestKeys)
         {
-            dataVersion = dataVer;
+            string calculatedClientDigest = CalculateDigest(digest, route, context.InputStream, auth, pspVersionInfo, isUpload, true);
+
+            // If the calculated client digest is invalid, then this isn't the digest the game is using, so check the next one
+            if (calculatedClientDigest != clientDigest) 
+                continue;
+            
+            SetDigestResponse(context, CalculateDigest(digest, route, context.ResponseStream, auth, null, isUpload, true));
+            
+            if(token != null)
+                gameDatabase.SetTokenDigestInfo(token, digest, true);
+            
+            return; 
```

Comment:
This code duplication can definitely be worked around. Something like `FindBestKey(..., bool isHmac)`

Reply:
i'm not entirely happy with how i solved this, but i couldn't figure out a cleaner solution given C#'s limitations